src/H5Dvirtual.c

```diff
@@ -3672,3 +3792,254 @@ H5D__virtual_release_source_dset_files(H5D_virtual_held_file_t *head)
+ *              num_mappings: Number of mappings in the array
+ *              leaves_out: Pointer to array of leaves, one per mapping that should be inserted
+ *                          Allocated on success and must be freed by caller.
+ *              is_in_tree_out: Pointer to boolean array, one per mapping, indicating
+ *                              whether the mapping is in the list of leaves.
+ *                              Allocated on success and must be freed by caller.
+ *              leaf_count: Pointer to number of leaves allocated in leaves_out.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__mappings_to_leaves(H5O_storage_virtual_ent_t *mappings, size_t num_mappings, H5RT_leaf_t **leaves_out,
+                        bool **is_in_tree_out, size_t *leaf_count)
+{
+    herr_t ret_value = SUCCEED;
+
+    H5RT_leaf_t *leaves_temp = NULL;
+    bool        *is_in_tree  = NULL;
+
+    bool should_insert_space = false;
+
+    H5O_storage_virtual_ent_t *curr_mapping    = NULL;
+    H5RT_leaf_t               *curr_leaf       = NULL;
+    size_t                     curr_leaf_count = 0;
+    H5S_t                     *curr_space      = NULL;
+
+    int rank = 0;
+
+    FUNC_ENTER_PACKAGE
+
+    assert(mappings);
+    assert(num_mappings > 0);
+    assert(leaf_count);
+    assert(leaves_out);
+    assert(is_in_tree_out);
+
+    /* Get rank from the first mapping's virtual selection */
+    if ((curr_space = mappings[0].source_dset.virtual_select) == NULL)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "first mapping has no virtual space");
+
+    if ((rank = H5S_GET_EXTENT_NDIMS(curr_space)) < 0 || rank > H5S_MAX_RANK)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get rank of dataspace");
+
+    if (rank == 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_BADVALUE, FAIL, "mapping has zero-dimensional space");
+
+    /* Allocate array of leaf structures */
+    if ((leaves_temp = (H5RT_leaf_t *)calloc(num_mappings, sizeof(H5RT_leaf_t))) == NULL)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "can't allocate leaves array");
+
+    if ((is_in_tree = (bool *)H5MM_calloc(num_mappings * sizeof(bool))) == NULL)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "can't allocate is_in_tree array");
+
+    for (size_t i = 0; i < num_mappings; i++) {
+        curr_mapping = &mappings[i];
+
+        if (H5D__rtree_should_insert((void *)curr_mapping, &should_insert_space) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTINSERT, FAIL, "error checking if mapping should be inserted");
+
+        if (!should_insert_space)
+            continue;
+
+        is_in_tree[i] = true;
+
+        /* Initialize leaf with dynamic coordinate allocation */
+        curr_leaf = &leaves_temp[curr_leaf_count];
+        if (H5RT_leaf_init(curr_leaf, rank, (uintptr_t)i) < 0)
```

Comment:
I'm still pretty sure you could have the r-tree use a void * for records - just pass `curr_mapping` instead of `i` here, and `curr_leaf->record` instead of `&storage->list[mapping_index]` in the invocation of H5D__virtual_pre_io_process_mapping() in the r-tree block of H5D__virtual_pre_io()